internal/controller/util/json_util.go

```diff
@@ -0,0 +1,348 @@
+package util
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/ramendr/ramen/internal/controller/kubeobjects"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/jsonpath"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func EvaluateCheckHook(client client.Client, hook *kubeobjects.HookSpec, log logr.Logger) (bool, error) {
+	timeout := getTimeoutValue(hook)
+	nsScopedName := types.NamespacedName{
+		Namespace: hook.Namespace,
+		Name:      hook.NameSelector,
+	}
+	switch hook.SelectResource {
+	case "pod":
+		// handle pod type
+		resource := &corev1.Pod{}
+		err := WaitUntilResourceExists(client, resource, nsScopedName, time.Duration(timeout)*time.Second)
+		if err != nil {
+			return false, err
+		}
+		podBytes, err := json.Marshal(resource)
+		if err != nil {
+			return false, err
+		}
+		jsonData := make(map[string]interface{})
+		err = json.Unmarshal(podBytes, &jsonData)
+		if err != nil {
+			return false, err
+		}
+		return evaluateCheckHookExp(hook.Chk.Condition, resource)
+	case "deployment":
+		// handle deployment type
+		resource := &appsv1.Deployment{}
+		err := WaitUntilResourceExists(client, resource, nsScopedName, time.Duration(timeout)*time.Second)
+		if err != nil {
+			return false, err
+		}
+		depBytes, err := json.Marshal(resource)
+		if err != nil {
+			return false, err
+		}
+		jsonData := make(map[string]interface{})
+		err = json.Unmarshal(depBytes, &jsonData)
+		if err != nil {
+			return false, err
+		}
+		return evaluateCheckHookExp(hook.Chk.Condition, resource)
+	case "statefulset":
+		// handle statefulset type
+		resource := &appsv1.StatefulSet{}
+		err := WaitUntilResourceExists(client, resource, nsScopedName, time.Duration(timeout)*time.Second)
+		if err != nil {
+			return false, err
+		}
+		statefulSetBytes, err := json.Marshal(resource)
+		if err != nil {
+			return false, err
+		}
+		jsonData := make(map[string]interface{})
+		err = json.Unmarshal(statefulSetBytes, &jsonData)
+		if err != nil {
+			return false, err
+		}
+		return evaluateCheckHookExp(hook.Chk.Condition, resource)
+	}
+
```

Comment:
All the cases have the same code, except the type of the resource. Maybe it can be combined?